Added OptionalOrElseGetConstant error prone check.

aioobe opened the pull request:
## Before this PR
Lots of `Optional.orElseGet(() -> someConstant)` floating around in the world.

## After this PR
==COMMIT_MSG==
Recommends using optional.orElse(constant) instead of optional.orElseGet(() -> constant)
==COMMIT_MSG==

## Possible downsides?
None

aioobe committed:
Added OptionalOrElseGetConstant error prone check.

aioobe committed:
Added OptionalOrElseGetConstant error prone check.

svc-excavator-bot committed:
Excavator: Upgrade buildscript dependencies (#1402)

Co-authored-by: svc-excavator-bot <svc-excavator-bot@palantir.com>

ferozco commented on baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetConstant.java:
```diff
@@ -0,0 +1,88 @@
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = SeverityLevel.ERROR,
+        summary = "If lambda passed to Optional#orElseGet returns a compile time constant, use Optional#orElse instead")
```
I think we should expand this check to include any variable

carterkozak replied:
+1 no need for it to be constant, field loads are less expensive than lambda allocations.

We would want to accept `foo`, `this.foo`, `foo.bar` but not `foo().bar`, `foo[index]`, `foo[index].bar`.

aioobe replied:
Was a bit worried about race conditions in case it's a member variable. Thinking of it now again though, I'm realizing that it's probably not an issue. I'll look into it.

aioobe replied:
Out of curiosity, why not `foo[index]` and `foo[index].bar`?

carterkozak replied:
I'm not entirely sure they'll always be better, but they probably are. Even if there's llc miss it's probably better than the lambda.
I tend to prefer flagging and fixing the most unambiguous cases, but I don't have a strong opinion here.

aioobe replied:
Check now covers `foo`, `this.foo`, `foo.bar` but not array accesses or anything that could have side-effects.

svc-excavator-bot committed:
Excavator: Upgrade buildscript dependencies (#1403)

Co-authored-by: svc-excavator-bot <svc-excavator-bot@palantir.com>

carterkozak commented on baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetConstant.java:
```diff
@@ -0,0 +1,88 @@
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "OptionalOrElseGetConstant",
```
If the suggested changes make sense, perhaps we should rename this to `OptionalOrElseGetValue`

carterkozak replied:
This check should be documented with the others in the readme

aioobe replied:
Added a line in the readme.

carterkozak commented on baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetConstant.java:
```diff
@@ -0,0 +1,88 @@
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = SeverityLevel.ERROR,
```
Should probably be `WARN` rather than `ERROR`.

aioobe replied:
Changed.

carterkozak commented on baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetConstant.java:
```diff
@@ -0,0 +1,88 @@
+        name = "OptionalOrElseGetConstant",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
```
After we test the suggested fixes on a large internal project (per-merge) we should add this check to the list that baseline automatically fixes by referencing the name from the list in `BaselineErrorProneExtension.java`

aioobe replied:
Sounds good. :+1:

iamdanfox committed:
--write-locks no longer implicitly runs checkClassUniqueness task (#1389)

Running `./gradlew --write-locks` should be faster now, as it doesn't update your `baseline-class-uniqueness.lock` file implicitly anymore. To update this, you need to run `./gradlew checkClassUniqueness --write-locks`.

svc-autorelease committed:
Autorelease 3.24.0

svc-excavator-bot committed:
Excavator: Upgrade dependencies (#1399)

* Excavator: Upgrade dependencies

* correct task

Co-authored-by: svc-excavator-bot <svc-excavator-bot@palantir.com>
Co-authored-by: Robert Kruszewski <robertk@palantir.com>

robert3005 committed:
Improvement: Upgrade Checkstyle to 8.33 (#1404)

Upgrade checkstyle to 8.33

svc-autorelease committed:
Autorelease 3.25.0

robert3005 committed:
Fix: Adjust checkstyle config for new version (#1409)

Fix checkstyle config to conform to breaks made between 8.13 and 8.33

svc-autorelease committed:
Autorelease 3.25.1

svc-excavator-bot committed:
Excavator: Render CircleCI file using template specified in .circleci/template.sh (#1384)

Co-authored-by: svc-excavator-bot <svc-excavator-bot@palantir.com>
Co-authored-by: forozco <forozco@palantir.com>

aioobe committed:
Check now also covers other simple expressions. Readme updated.

aioobe committed:
Check now also covers other simple expressions. Readme updated.

carterkozak commented on README.md:
```diff
@@ -159,6 +159,7 @@ Safe Logging can be found at [github.com/palantir/safe-logging](https://github.c
 - `NonComparableStreamSort`: Stream.sorted() should only be called on streams of Comparable types.
 - `DangerousStringInternUsage`: Disallow String.intern() invocations in favor of more predictable, scalable alternatives.
 - `OptionalOrElseThrowThrows`: Optional.orElseThrow argument must return an exception, not throw one.
+- `OptionalOrElseGetValue`: Prefer `Optional.orElse(expression)` over `Optional.orElseGet(() -> expression)` for trivial expressions.
```
pedantic nit:
```suggestion
- `OptionalOrElseGetValue`: Prefer `Optional.orElse(value)` over `Optional.orElseGet(() -> value)` for trivial expressions.
```

carterkozak commented on baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetValue.java:
```diff
@@ -0,0 +1,99 @@
+            }
+
+            ExpressionTree expressionBody = (ExpressionTree) lambdaExpressionTree.getBody();
+            if (COMPILE_TIME_CONSTANT.matches(expressionBody, state)
```
Is the compile time constant check ever necessary? What does it catch that the following checks cannot?

aioobe replied:
It covers expressions that may look more complicated than the other cases, but can be evaluated at compile time. For example `PREFIX + "_someSuffix"` or `MIN + SIZE * 1.5`.

Added a test for such case.

ferozco committed:
StrictUnusedVariable handles Java 14 records (#1412)

StrictUnusedVariable handles Java 14 records

svc-autorelease committed:
Autorelease 3.26.0

aioobe committed:
Ran spotlessApply.

aioobe committed:
Ran spotlessApply.

aioobe committed:
Added a compile time constant test.

aioobe committed:
Added a compile time constant test.

aioobe committed:
Reworded description in README.

aioobe committed:
Reworded description in README.

carterkozak commented on baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetValue.java:
```diff
@@ -0,0 +1,98 @@
+            LambdaExpressionTree.BodyKind bodyKind = lambdaExpressionTree.getBodyKind();
+
+            if (bodyKind != LambdaExpressionTree.BodyKind.EXPRESSION) {
+                break match;
```
Not sure the named block + break buys us anything over `return Description.NO_MATCH;`, by returning no-match we can shed a layer of indents, and style more uniformly with other checks.

aioobe replied:
It was part of my battle against cyclomatic complexity. I'll check if I can drop it without tipping over.

carterkozak replied:
I’d suppress the cyclomatic complexity check here

aioobe committed:
Changed to multiple returns.

aioobe committed:
Changed to multiple returns.

hesselink committed:
Don't wrap method declaration (#1416)

Wrap method declarations on parameters, not on modifiers and return type.

svc-autorelease committed:
Autorelease 3.27.0

svc-excavator-bot committed:
Excavator: Upgrade buildscript dependencies (#1417)

Co-authored-by: svc-excavator-bot <svc-excavator-bot@palantir.com>

svc-excavator-bot committed:
Excavator: Ensure consistent gradle/publish-(jar|dist).gradle files (#1418)

Co-authored-by: svc-excavator-bot <svc-excavator-bot@palantir.com>

svc-excavator-bot committed:
Excavator: Upgrade dependencies (#1419)

Co-authored-by: svc-excavator-bot <svc-excavator-bot@palantir.com>

aioobe committed:
Added changelog entry.

aioobe committed:
Added changelog entry.

aldexis committed:
Improve gradle-baseline-java integration with IntelliJ import (#1411)

Adds the proper configuration files upon IntelliJ import of a gradle project for checkstyle and copyright.

This generates the following additional files:
- .idea/copyright/profiles_settings.xml
- an xml file under .idea/copyright/ per copyright file under .baseline/copyright
- .idea/checkstyle-idea.xml (and adds Checkstyle-IDEA to the external dependencies) if baseline-checkstyle is applied
- Either .idea/codeStyleSettings.xml or a .idea/codeStyles/ folder with the contents being copied from .baseline/idea
  - If .baseline/idea/codeStyles is present, it will copy its contents, otherwise, it will fall back to .baseline/idea/intellij-java-palantir-style.xml as currently
  - The fallback is using a legacy IntelliJ format and requires closing and reopening the project to be taken into account

svc-autorelease committed:
Autorelease 3.28.0

carterkozak commented:
do we need to merge develop into this branch to pick up the unit-test-14 checks?

aioobe committed:
Merge branch 'orElseGet-constant' of github.com:aioobe/gradle-baseline into orElseGet-constant
